scalding-core/src/main/scala/com/twitter/scalding/estimation/Estimator.scala

```diff
@@ -0,0 +1,67 @@
+package com.twitter.scalding.estimation
+
+import cascading.flow.{ Flow, FlowStep }
+import com.twitter.algebird.Monoid
+import org.apache.hadoop.mapred.JobConf
+import org.slf4j.LoggerFactory
+import scala.util.{ Failure, Success }
+
+case class FlowStrategyInfo(
+  flow: Flow[JobConf],
+  predecessorSteps: Seq[FlowStep[JobConf]],
+  step: FlowStep[JobConf])
+
+/**
+ * Trait for estimation some parameters of Job.
+ * @tparam T return type of estimation
+ */
+trait Estimator[T] {
+  def estimate(info: FlowStrategyInfo): Option[T]
+}
+
+case class FallbackEstimator[T](first: Estimator[T], fallback: Estimator[T]) extends Estimator[T] {
+  private val LOG = LoggerFactory.getLogger(this.getClass)
+
+  override def estimate(info: FlowStrategyInfo): Option[T] = {
+    first.estimate(info).orElse {
+      LOG.warn(s"$first estimator failed. Falling back to $fallback.")
+      fallback.estimate(info)
+    }
+  }
+}
+
+class FallbackEstimatorMonoid[T] extends Monoid[Estimator[T]] {
+  override def zero: Estimator[T] = new Estimator[T] {
+    override def estimate(info: FlowStrategyInfo): Option[T] = None
+  }
+
+  override def plus(l: Estimator[T], r: Estimator[T]): Estimator[T] = FallbackEstimator(l, r)
+}
+
+trait HistoryEstimator[T] extends Estimator[T] {
+  private val LOG = LoggerFactory.getLogger(this.getClass)
+
+  def maxHistory(conf: JobConf): Int
+
+  def historyService: HistoryService
+
+  override def estimate(info: FlowStrategyInfo): Option[T] = {
+    val conf = info.step.getConfig
+
+    historyService.fetchHistory(info, maxHistory(conf)) match {
+      case Success(history) if history.isEmpty =>
+        LOG.warn("No matching history found.")
```

Comment:
How about `s"No matching history found for $info"` or something? In an execution, I can imagine this getting confusing.